storage/inmem/ast.go

```diff
@@ -0,0 +1,313 @@
+		}
+
+		switch op {
+		case storage.AddOp:
+			var results []*ast.Term
+			for i := 0; i < data.Len(); i++ {
+				if i == pos {
+					results = append(results, ast.NewTerm(value))
+				}
+				results = append(results, data.Elem(i))
+			}
+
+			return &updateAST{path[:len(path)-1], false, ast.NewArray(results...)}, nil
+
+		case storage.RemoveOp:
+			var results []*ast.Term
+			for i := 0; i < data.Len(); i++ {
+				if i != pos {
+					results = append(results, data.Elem(i))
+				}
+			}
+			return &updateAST{path[:len(path)-1], false, ast.NewArray(results...)}, nil
+
+		default:
+			var results []*ast.Term
+			for i := 0; i < data.Len(); i++ {
+				if i == pos {
+					results = append(results, ast.NewTerm(value))
+				} else {
+					results = append(results, data.Elem(i))
+				}
+			}
+
+			return &updateAST{path[:len(path)-1], false, ast.NewArray(results...)}, nil
+		}
+	}
+
+	pos, err := ptr.ValidateASTArrayIndex(data, path[idx], path)
+	if err != nil {
+		return nil, err
+	}
+
+	return newUpdateAST(data.Elem(pos).Value, op, path, idx+1, value)
+}
+
+func newUpdateObjectAST(data ast.Object, op storage.PatchOp, path storage.Path, idx int, value ast.Value) (*updateAST, error) {
+	key := ast.StringTerm(path[idx])
+	val := data.Get(key)
+
+	if idx == len(path)-1 {
+		switch op {
+		case storage.ReplaceOp, storage.RemoveOp:
+			if val == nil {
+				return nil, errors.NewNotFoundError(path)
+			}
+		}
+		return &updateAST{path, op == storage.RemoveOp, value}, nil
+	}
+
+	if val != nil {
+		return newUpdateAST(val.Value, op, path, idx+1, value)
+	}
+
+	return nil, errors.NewNotFoundError(path)
+}
+
+func interfaceToValue(v interface{}) (ast.Value, error) {
+	if v, ok := v.(ast.Value); ok {
+		return v, nil
+	}
+	return ast.InterfaceToValue(v)
+}
+
+// setInAst updates the value in the AST at the given path with the given value.
+// Values can only be replaced in arrays, not added.
+// Values for new keys can be added to objects
+func setInAst(data ast.Value, path storage.Path, value ast.Value) (ast.Value, error) {
+	if len(path) == 0 {
+		return data, nil
+	}
+
+	switch data := data.(type) {
+	case ast.Object:
+		return setInAstObject(data, path, value)
+	case *ast.Array:
+		return setInAstArray(data, path, value)
+	default:
+		return nil, fmt.Errorf("illegal value type %T, expected ast.Object or ast.Array", data)
+	}
+}
+
+func setInAstObject(obj ast.Object, path storage.Path, value ast.Value) (ast.Value, error) {
+	key := ast.StringTerm(path[0])
+
+	if len(path) == 1 {
+		obj.Insert(key, ast.NewTerm(value))
+		return obj, nil
+	}
+
+	child := obj.Get(key)
+	newChild, err := setInAst(child.Value, path[1:], value)
+	if err != nil {
+		return nil, err
+	}
+	obj.Insert(key, ast.NewTerm(newChild))
+	return obj, nil
+}
+
+func setInAstArray(arr *ast.Array, path storage.Path, value ast.Value) (ast.Value, error) {
+	idx, err := strconv.Atoi(path[0])
+	if err != nil {
+		// We expect the path to be valid at this point.
+		return arr, nil
```

Comment:
I might have missed this in a refactoring. Since we've already asserted a correct path when creating the update (`store.newUpdate()`), I don't think this has any real consequence - which is why we have no failing tests. But I'll fix this to be consistent 👍.